code/datums/browser.dm

```diff
@@ -100,11 +100,107 @@
 		window_size = "size=[width]x[height];"
 	user << browse(get_content(), "window=[window_id];[window_size][window_options]")
 	if (use_onclose)
-		onclose(user, window_id, ref)
+		spawn(0)
+			//winexists sleeps, so we don't need to.
+			for (var/i = 0, !winexists(user, window_id) && i < 3, i++)
+				onclose(user, window_id, ref)
+
 
 /datum/browser/proc/close()
 	user << browse(null, "window=[window_id]")
 
+/datum/browser/alert
+	var/selectedbutton = 0
+	var/opentime = 0
+	var/timeout
+	var/stealfocus
+
+/datum/browser/alert/New(User,Message,Title,Button1="Ok",Button2,Button3,StealFocus = 1,Timeout=6000)
+	if (!User)
+		return
+	var/output =  {"<center><b>[Message]</b></center><br />
+		<a style="font-size:large;float:left" href="?src=\ref[src];button=1">[Button1]</a>"}
+
+	if (Button2)
+		output += {"<a style="font-size:large;float:right" href="?src=\ref[src];button=2">[Button2]</a>"}
+
+	if (Button3)
+		EXCEPTION("button 3 is not impimented yet")
+
+
+	..(User, ckey("[User]-[Message]-[Title]-[world.time]-[rand(1,10000)]"), Title, 350, 150, src)
+	set_content(output)
+	stealfocus = StealFocus
+	if (!StealFocus)
+		window_options += "focus=false;"
+	timeout = Timeout
+
+/datum/browser/alert/open()
+	set waitfor = 0
+	opentime = world.time
+
+	if (stealfocus)
+		. = ..(use_onclose = 1)
+	else
+		var/focusedwindow = winget(user, null, "focus")
+		. = ..(use_onclose = 1)
+
+		//waits for the window to show up client side before attempting to un-focus it
+		//winexists sleeps until it gets a reply from the client, so we don't need to bother sleeping
+		for (var/i = 0, user && !winexists(user, window_id) && i < 10, i++)

```

Comment:
if the part of the loopheader that's an expression (the middle bit) is false, it stops (as in the break command)